core/services/ocr2/plugins/ocr2keeper/evm21/upkeepstate/store.go

```diff
@@ -200,13 +214,34 @@ func (u *upkeepStateStore) upsertStateRecord(ctx context.Context, workID string,
 
 	u.cache[workID] = record
 
-	return u.orm.InsertUpkeepState(persistedStateRecord{
+	u.pendingRecords = append(u.pendingRecords, persistedStateRecord{
 		UpkeepID:            utils.NewBig(upkeepID),
 		WorkID:              record.workID,
 		CompletionState:     uint8(record.state),
 		IneligibilityReason: reason,
 		InsertedAt:          record.addedAt,
-	}, pg.WithParentCtx(ctx))
+	})
+
+	if len(u.pendingRecords) >= flushSize {
+		u.flush(ctx)
+	}
```

Comment:
as discussed offline, might be cleaner to make flush independent of upsertRecords and trigger it on a timer. (while flushing records in batches)